### ISSUE TITLE
[PyTorch Upstream][B60] 1 confirmed performance drops with new triton 

### ISSUE THREAD

**Issue by mengfei25** [intel/intel-xpu-backend-for-triton#8077]
## Summary
Triton-git00b9d1ae validation on BMG shows 1 confirmed drops compared with Triton-git1e2d42a0 test after applying rerun verification.

## Affected case
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kaileiyx/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kaileiyx/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">

Suite | Dtype | Mode | Model | Ind Ratio | Eager Ratio
-- | -- | -- | -- | -- | --
timm_models | bfloat16 | training | mixnet_l | 0.8626 | 0.9973
timm_models | float16 | training | mixnet_l | 0.857 | 0.9971

</body>

</html>

## Reproducer
```bash
docker run -it -e TZ=Asia/Shanghai \
  --device=/dev/mem --device=/dev/dri --group-add video \
  --privileged --shm-size=8g \
  intelgpu/ubuntu-26.04-rolling:26.18 bash

##Target triton
gh --repo pytorch/pytorch run download 34471485331 -p "triton-wheel-3.10-xpu-manylinux_2_28_x86_64"
 
curl -LsSf https://astral.sh/uv/install.sh | env UV_INSTALL_DIR="/usr/local/bin" sh
uv venv myvenv --python 3.10 --clear
source myvenv/bin/activate
uv pip install pip wheel "setuptools<81"

uv pip install torch==2.15.0.dev20260912+xpu torchvision==0.30.0.dev20260913+xpu torchaudio==2.11.0.dev20260913+xpu triton-xpu==3.8.0+git1e2d42a0 --pre --index-url https://download.pytorch.org/whl/nightly/xpu

pytorch_commit="$(python -c 'import torch; print(torch.version.git_version)')"
git clone https://github.com/pytorch/pytorch
cd pytorch
git checkout ${pytorch_commit}

uv pip install pandas psutil scipy requests
uv pip install -U numpy==1.26.4 accelerate
uv pip install -r .ci/docker/ci_commit_pins/huggingface-requirements.txt
uv pip install timm

git clone https://github.com/intel/torch-xpu-ops
rsync -avz ./torch-xpu-ops/.ci/benchmarks/ ./benchmarks/dynamo/
##Install target triton
uv pip uninstall triton-xpu -y
uv pip install /download/path/triton-xpu_*.wheel
##run command line
python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --training --float16 --only mixnet_l
python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --training --bfloat16 --only mixnet_l
```

### Versions

torch                   2.15.0.dev20260912+xpu
torchao                 0.19.0.dev20260913+xpu
torchaudio              2.11.0.dev20260913+xpu
torchvision             0.30.0.dev20260913+xpu
triton-xpu              3.8.0+git00b9d1ae

**Comment by anmyachev** [comment]
Hi @mengfei25, what is the size of the regression? In our CI we have [1.293x](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/34953163535/job/104328790611#step:23:95) (release 3.8.0) vs [1.244x](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/34953089079/job/104328546654#step:23:97) (release 3.9.0).

PyTorch ref: 23b1588b32ea2d46410793347d635017fdc49c75


**UPD:** with new pytorch pin we have [1.042x](https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/35131690954/job/104914038092#step:23:97).

PyTorch ref: 76872935231bf7ad36cd1923a68c58572bc57c36

At the moment, it seems like regression in PyTorch. @mengfei25 could you recheck?

**Comment by mengfei25** [comment]
@anmyachev 
We need to extract the absolute Eager and Inductor latencies from the output CSV. Since the printed speedup ratio is simply calculated as Eager/Inductor, relying on it alone is not accurate enough to detect a pure Eager or Inductor performance regression.

In this case, we are looking at a pure Inductor performance drop (inductor slower ~15%). You can find the absolute latency data `inductor = abs_latency; eager = abs_latency * speedup` inside the output CSV by passing the `--output /path/to/xxx.csv` argument.

Additionally, all our tests were conducted using same env, `torch==2.15.0.dev20260912+xpu`, with the only variable being the Triton commit. Therefore, we strongly suspect this is a Triton-side regression.

Could you please help verify if this is reproducible on your end? If the investigation confirms it is actually a PyTorch-core regression instead, we will gladly take ownership of it.

**Comment by anmyachev** [comment]
I've found the following (on PVC + rolling):
Bad:
```bash
dev,name,batch_size,speedup,abs_latency,compilation_latency,compression_ratio,eager_peak_mem,dynamo_peak_mem,calls_captured,unique_graphs,fallbacks_to_eager,graph_breaks,unique_graph_breaks,autograd_captures,autograd_compiles,cudagraph_skips
xpu,mixnet_l,128,0.800327,261.961601,183.497898,0.000000,0,0,579,1,0,6,5,0,0,1
```
good:
```bash
dev,name,batch_size,speedup,abs_latency,compilation_latency,compression_ratio,eager_peak_mem,dynamo_peak_mem,calls_captured,unique_graphs,fallbacks_to_eager,graph_breaks,unique_graph_breaks,autograd_captures,autograd_compiles,cudagraph_skips
xpu,mixnet_l,128,1.083409,193.259441,183.253093,0.000000,0,0,579,1,0,6,5,0,0,1
```

and I've bisected to:
```bash
cdd9908bc7d28b581186d3279259e093241c0210 is the first bad commit
commit cdd9908bc7d28b581186d3279259e093241c0210
Author: Animesh Jain <anijain@umich.edu>
Date:   Mon Aug 31 13:23:56 2026 -0700

    [functionalization] Regenerate one output of a multi-output view without replaying the op (#194037)
    
    Functionalization records a ViewMeta per view so it can rebuild that view after
    its base is mutated. For a multi-output view the recorded replay is the whole
    operation plus an index, so one output of unbind is rebuilt as
    `at::unbind(base, dim)[i]`. Rebuilding all N outputs of one base allocates N*N
    tensors, and under torch.compile N*N fake tensors and proxies, leaving one
    redundant split node in the post-AOT graph per regeneration.
```

I am waiting for the B60 machine to reproduce in exactly env you reported. There might be a different reason here. cc @kwasd.

**Comment by mengfei25** [comment]
> I've found the following (on PVC + rolling): Bad:
> 
> dev,name,batch_size,speedup,abs_latency,compilation_latency,compression_ratio,eager_peak_mem,dynamo_peak_mem,calls_captured,unique_graphs,fallbacks_to_eager,graph_breaks,unique_graph_breaks,autograd_captures,autograd_compiles,cudagraph_skips
> xpu,mixnet_l,128,0.800327,261.961601,183.497898,0.000000,0,0,579,1,0,6,5,0,0,1
> good:
> 
> dev,name,batch_size,speedup,abs_latency,compilation_latency,compression_ratio,eager_peak_mem,dynamo_peak_mem,calls_captured,unique_graphs,fallbacks_to_eager,graph_breaks,unique_graph_breaks,autograd_captures,autograd_compiles,cudagraph_skips
> xpu,mixnet_l,128,1.083409,193.259441,183.253093,0.000000,0,0,579,1,0,6,5,0,0,1
> and I've bisected to:
> 
> cdd9908bc7d28b581186d3279259e093241c0210 is the first bad commit
> commit cdd9908bc7d28b581186d3279259e093241c0210
> Author: Animesh Jain <anijain@umich.edu>
> Date:   Mon Aug 31 13:23:56 2026 -0700
> 
>     [functionalization] Regenerate one output of a multi-output view without replaying the op (#194037)
>     
>     Functionalization records a ViewMeta per view so it can rebuild that view after
>     its base is mutated. For a multi-output view the recorded replay is the whole
>     operation plus an index, so one output of unbind is rebuilt as
>     `at::unbind(base, dim)[i]`. Rebuilding all N outputs of one base allocates N*N
>     tensors, and under torch.compile N*N fake tensors and proxies, leaving one
>     redundant split node in the post-AOT graph per regeneration.
> I am waiting for the B60 machine to reproduce in exactly env you reported. There might be a different reason here. cc [@kwasd](https://github.com/kwasd).

We also found some perf drops on PVC, tracked by https://github.com/intel/intel-xpu-backend-for-triton/issues/8106

**Comment by anmyachev** [comment]
Ok, was able to reproduce in that docker container: `1.211x` vs `1.043x`. IGC in our CI: `LIBIGC1_VERSION=2.40.13` and IGC in the docker container: `LIBIGC1_VERSION=2.34.4`. That might be the reason. Thanks @chuanqi129 for the machine!

**Comment by anmyachev** [comment]
Regression:
```bash
git bisect good
00b9d1aef6ec45fe29b5d16052268f02523f5ce0 is the first bad commit
commit 00b9d1aef6ec45fe29b5d16052268f02523f5ce0
Author: Ettore Tiotto <ettore.tiotto@intel.com>
Date:   Wed Sep 9 07:53:21 2026 -0400

    [XPU] Gate the auto-large-GRF retry on the reported n_spills (#7959)

    The auto-large-GRF rebuild recompiles a kernel at 256 GRF whenever the
    128-GRF build spills more than a threshold. That threshold was a flat 0
    bytes (#7615), so any spill at all triggered a recompile.
```

Command to reproduce: `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 python benchmarks/dynamo/timm_models.py --performance -d xpu -n10 --backend=inductor --cold-start-latency --training --float16 --only mixnet_l`

@etiotto the same regression as in https://github.com/intel/intel-xpu-backend-for-triton/issues/8106. Please take a look. Note: it's reproducible only for `LIBIGC1_VERSION=2.34.4`.

**Comment by anmyachev** [comment]
@etiotto just FYI: reducing `MAX_REG_SPILL_SLOTS_PER_LANE` to 8 helps here.

**Comment by etiotto** [comment]
@anmyachev, @mengfei25: Unless the regression is reproducible on the latest rolling, this is a moot point right? So the important factor is is whether the workload exhibits a regression on the latest rolling driver.

**Comment by anmyachev** [comment]
> [@anmyachev](https://github.com/anmyachev), [@mengfei25](https://github.com/mengfei25): Unless the regression is reproducible on the latest rolling, this is a moot point right? So the important factor is is whether the workload exhibits a regression on the latest rolling driver.

@etiotto It's not reproducible on IGC 2.40 (see https://github.com/intel/intel-xpu-backend-for-triton/issues/8077#issuecomment-5695610987)

I assume the current situation is such that the change - while causing problems in that particular case - generally improves more cases, right @etiotto? @mengfei25 @chuanqi129 What about leaving everything as is and just waiting for the new driver version in PyTorch?

**Comment by etiotto** [comment]
@anmyachev I have a PR to fix this regression for the release here: https://github.com/intel/intel-xpu-backend-for-triton/pull/8148
